src/dotnet/commands/dotnet-build/IncrementalManager.cs

```diff
@@ -0,0 +1,201 @@
+            if (result.NeedsRebuilding)
+            {
+                return result;
+            }
+
+            result = InputItemsChanged(graphNode, compilerIO);
+            if (result.NeedsRebuilding)
+            {
+                return result;
+            }
+
+            result = TimestampsChanged(compilerIO);
+            if (result.NeedsRebuilding)
+            {
+                return result;
+            }
+
+            return IncrementalResult.DoesNotNeedRebuild;
+            
+        }
+
+        private IncrementalResult CLIChanged(ProjectGraphNode graphNode)
+        {
+            var currentVersionFile = DotnetFiles.VersionFile;
+            var versionFileFromLastCompile = graphNode.ProjectContext.GetSDKVersionFile(_configuration, _buildBasePath, _outputPath);
+
+            if (!File.Exists(currentVersionFile))
+            {
+                // this CLI does not have a version file; cannot tell if CLI changed
+                return IncrementalResult.DoesNotNeedRebuild;
+            }
+
+            if (!File.Exists(versionFileFromLastCompile))
+            {
+                // this is the first compilation; cannot tell if CLI changed
+                return IncrementalResult.DoesNotNeedRebuild;
+            }
+
+            var currentContent = DotnetFiles.ReadAndInterpretVersionFile();
+
+            var versionsAreEqual = string.Equals(currentContent, File.ReadAllText(versionFileFromLastCompile), StringComparison.OrdinalIgnoreCase);
+
+            return versionsAreEqual
+                ? IncrementalResult.DoesNotNeedRebuild
+                : new IncrementalResult("the version or bitness of the CLI changed since the last build");
+        }
+
+        private IncrementalResult InputItemsChanged(ProjectGraphNode graphNode, CompilerIO compilerIO)
+        {
+            // check empty inputs / outputs
+            if (!compilerIO.Inputs.Any())
+            {
+                return new IncrementalResult("the project has no inputs");
+            }
+
+            if (!compilerIO.Outputs.Any())
+            {
+                return new IncrementalResult("the project has no outputs");
+            }
+
+            // check non existent items
+            var result = CheckMissingIO(compilerIO.Inputs, "inputs");
+            if (result.NeedsRebuilding)
+            {
+                return result;
+            }
+
+            result = CheckMissingIO(compilerIO.Outputs, "outputs");
+            if (result.NeedsRebuilding)
+            {
+                return result;
+            }
+
+            // check cache against input glob pattern changes
+            var incrementalCacheFile = graphNode.ProjectContext.IncrementalCacheFile(_configuration, _buildBasePath, _outputPath);
+
+            if (!File.Exists(incrementalCacheFile))
+            {
+                // no cache present; cannot tell if anything changed
+                return IncrementalResult.DoesNotNeedRebuild;
```

Comment:
?? If we don't find a cache file we say that we don't need to rebuild?  I would think it would be the opposite.